### PR TITLE
sourceLocations is broken when tags are implied

### PR DESCRIPTION
**Pull request by thewilkybarkid**
This is deliberately left as a draft as it should demonstrate a bug as a result of https://github.com/fb55/htmlparser2/issues/896, which causes https://github.com/parcel-bundler/parcel/issues/6672. 

I don't think it's possible to work around it here, but there will probably need to be code changes when the htmlparser2 bug is resolved.

### CODE REVIEW

**Comment by thewilkybarkid** [comment]
htmlparser2 has had a code change that looks to correct the issue there (https://github.com/fb55/htmlparser2/pull/910), but the problem is still seen here (as the same indices are seen twice).

File: src/index.ts [hunk 1]
```diff
@@ -143,7 +143,7 @@ export const parser = (html: string, options: Options = {}): Node[] => {
     const buf: Node | undefined = bufArray.pop();
 
     if (buf && typeof buf === 'object' && buf.location && parser.endIndex !== null) {
-      buf.location.end = locationTracker.getPosition(parser.endIndex);
+      buf.location.end = locationTracker.getPosition((!parser.tagname || parser.tagname === buf.tag) ? parser.endIndex : parser.startIndex - 1);
```
Comment (thewilkybarkid) [inline src/index.ts:146]: At the moment I can't see a way to know if the closing tag is implied or not. It seems to be possible to determine it with `tagname`... but it's private.

Comment (thewilkybarkid) [reply src/index.ts:146]: The only other way I can think of is to see if an open tag has the same indices as the last close tag and rewrite it.

Comment (thewilkybarkid) [reply src/index.ts:146]: @fb55 Is there another way to know if the closing tag is implied?

Comment (fb55) [reply src/index.ts:146]: One easy way would be to see if the indices have changed between events. What complicates this is that there can also be implied opening tags. Eg. `</p>` without a paragraph on the stack will lead to a `<p>` tag to be opened.

As an aside: I have just published `htmlparser2@7.0.0`, which has a much saner way of calculating indices.

Comment (thewilkybarkid) [reply src/index.ts:146]: Thanks. Hadn't thought about invalid HTML, makes this even more fun. 😅

I'll take a look at trying to keep track of them.

**Comment by thewilkybarkid** [comment]
_Mostly_ works, but the start of the `</p>` paragraph seems to be calculated as the `/` rather than the `<`.

**Comment by thewilkybarkid** [comment]
Thanks, @fb55; unfortunately that change doesn't seem to make a difference...

**Comment by fb55** [comment]
> unfortunately that change doesn't seem to make a difference

The start of a closing tag is now set to the `<` instead of the `/`. Do you have an example where it doesn't work?

**Comment by thewilkybarkid** [comment]
Unfortunately, CI isn't running on this PR yet, but the test fails on the implied start tag for the `</p>`:

```
      {
        location: {
          end: Object { … },
          start: {
  -         column: 19,
  +         column: 18,
            line: 2,
          },
        },
        tag: 'p',
      },
```

with and without the change.

**Comment by fb55** [comment]
Good find! Turns out the index was overwritten for implied opening tags. I've opened https://github.com/fb55/htmlparser2/pull/917 with a fix.

**Comment by thewilkybarkid** [comment]
@fb55 Thanks, the test's now passing. 😃

File: src/location-tracker.ts [hunk 2]
```diff
@@ -25,7 +25,12 @@ export class LocationTracker {
 
   getPosition(index: number): Position {
     if (index < this.lastIndex) {
-      throw new Error('Source indices must be monotonic');
+      this.lastPosition = {
+        line: 1,
+        column: 1
+      };
+
+      this.lastIndex = 0;
```
Comment (thewilkybarkid) [inline src/location-tracker.ts:33]: There might be another solution to this, but the problem comes from https://github.com/posthtml/posthtml-parser/blob/41845391201256ec5e676c2d98ef292a80924b5c/src/index.ts#L138 which is easy to solve, except for when it needs to know the end of the previous line.

**Comment by fb55** [comment]
I just published `htmlparser2@7.1.0`. With this version you'll be able to check an `isImplied` flag on `onopentag`/`onclosetag` events, which should make this PR much simpler: https://github.com/fb55/htmlparser2/pull/930

**Comment by thewilkybarkid** [comment]
That's great, thanks!

**Comment by thewilkybarkid** [comment]
Marking this as ready as there's a stable htmlparser release that's working.

Had a quick go at using the `isImplied` options and it doesn't seem to work as expected, the `</b>` in the test seems to flagged as implied, when it's not. /cc @fb55

**Comment by fb55** [comment]
> Had a quick go at using the `isImplied` options and it doesn't seem to work as expected

Thanks for flagging! I published `7.1.1` with a fix for this, as well as for text events having wrong indices.

**Comment by thewilkybarkid** [comment]
Thanks, @fb55. It's now passing, and it is definitely simpler. ✨

**Review by Scrum — APPROVED** [review]

**Comment by Scrum** [comment]
@thewilkybarkid @fb55 Thanks guys, released in version 0.10.1

**Pull request merged**